xmodule/capa_block.py

```diff
@@ -2222,6 +2332,102 @@ def score_from_lcp(self, lcp):
         return Score(raw_earned=lcp_score['score'], raw_possible=lcp_score['total'])
 
 
+class GradingMethodHandler:
+    """
+    A class for handling grading method and calculating scores.
+
+    This class allows for flexible handling of grading methods, including options
+    such as considering the last score, the first score, the highest score,
+    or the average score.
+
+    Attributes:
+        - score (Score): The current score.
+        - grading_method (str): The chosen grading method.
+        - score_history (list[Score]): A list to store the history of scores.
+        - max_score (int): The maximum possible score.
+        - mapping_method (dict): A dictionary mapping the grading
+            method to the corresponding handler.
+
+    Methods:
+        - get_score(): Retrieves the updated score based on the grading method.
+        - handle_last_score(): Handles the last score method.
+        - handle_first_score(): Handles the first score method.
+        - handle_highest_score(): Handles the highest score method.
+        - handle_average_score(): Handles the average score method.
+    """
+
+    def __init__(
+        self,
+        score: Score,
+        grading_method: str,
+        score_history: list[Score],
+        max_score: int,
+    ):
+        self.score = score
+        self.grading_method = grading_method
+        self.score_history = score_history
+        self.max_score = max_score
+        self.mapping_method = {
+            GRADING_METHOD.LAST_SCORE: self.handle_last_score,
+            GRADING_METHOD.FIRST_SCORE: self.handle_first_score,
+            GRADING_METHOD.HIGHEST_SCORE: self.handle_highest_score,
+            GRADING_METHOD.AVERAGE_SCORE: self.handle_average_score,
+        }
+
+    def get_score(self) -> Score:
+        """
+        Retrieves the updated score based on the grading method.
+
+        Returns:
+            - Score: The updated score based on the chosen grading method.
+        """
+        return self.mapping_method[self.grading_method]()
```

Comment:
how's the default handled here?

Reply:
When the component is created the default value for the grading method is `last_score`. Therefore, upon receiving that value, the `self.handle_last_score` method would be called.

Reply:
Where does that happen exactly?

Reply:
It is by the `self.mapping_method`, which returns the corresponding method according to the value of `self.grading_method`

Reply:
But self.mapping_method is a simple dictionary, when not specifying a grading method (meaning we need to use the default), how is the default specified?

Reply:
The default value is taken from the `grading_method` field defined in [`ProblemXBlock`](https://github.com/openedx/edx-platform/blob/862b942363576e03973f9a21e201c3626847c91a/xmodule/capa_block.py#L202).